crates/pathfinder/src/sync/p2p/headers.rs

```diff
@@ -0,0 +1,116 @@
+use anyhow::Context;
+use pathfinder_common::{BlockCommitmentSignature, BlockHeader, BlockNumber};
+use pathfinder_storage::Storage;
+use tokio::sync::mpsc::{Receiver, Sender};
+
+const CHUNK_SIZE: usize = 1024;
+const CHANNEL_SIZE: usize = 2;
+
+pub async fn spawn(
+    storage: Storage,
+    p2p: (),
+    max: BlockNumber,
+    min: BlockNumber,
+) -> anyhow::Result<()> {
+    let (tx_download, rx_verify) = tokio::sync::mpsc::channel(CHANNEL_SIZE);
+    let (tx_verify, rx_persist) = tokio::sync::mpsc::channel(CHANNEL_SIZE);
+
+    let jh_download = tokio::spawn(download(tx_download, p2p, max, min, CHUNK_SIZE));
+    let jh_verify = std::thread::spawn(|| verify(rx_verify, tx_verify));
+    let jh_persist = std::thread::spawn(|| persist(rx_persist, storage));
+
+    // Since all stages are connected, we can simply wait for the final stage to complete.
+    // This works even in the event of an earlier stage failing for some reason since the
+    // channel closures will cascade and cause the final stage to exit once all data in the
+    // pipeline has been processed.
+    let result = tokio::task::spawn_blocking(|| jh_persist.join())
+        .await
+        .context("Awaiting persist thread")?;
+
+    // TODO: join the rest
+    // TODO: report any errors upstream
+
+    Ok(())
+}
+
+/// Downloads batches of block headers via p2p in reverse chronological order.
+async fn download(
+    dest: Sender<Vec<(BlockHeader, BlockCommitmentSignature)>>,
+    p2p: (), // p2p source / API
+    mut max: BlockNumber,
+    min: BlockNumber,
+    chunk_size: usize,
+) -> anyhow::Result<()> {
+    // TODO: include the hash of `max` so we know we are getting what we expect.
```

Comment:
peers can reply with a shorter range than requested (`Fin(too_long)`), which is not an error btw, so we need to remember to take this into account

Reply:
Yeah I don't see this as being this code's problem. What `p2p` is should handle all of that.